presto-main/src/test/java/com/facebook/presto/cost/TestJoinStatsRule.java

```diff
@@ -0,0 +1,218 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static com.facebook.presto.metadata.MetadataManager.createTestMetadataManager;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.LEFT;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+
+public class TestJoinStatsRule
+{
+    private static final String LEFT_JOIN_COLUMN = "left_join_column";
+    private static final String RIGHT_JOIN_COLUMN = "right_join_column";
+    private static final String LEFT_OTHER_COLUMN = "left_column";
+    private static final String RIGHT_OTHER_COLUMN = "right_column";
+
+    private static final double LEFT_ROWS_COUNT = 500.0;
+    private static final double RIGHT_ROWS_COUNT = 1000.0;
+    private static final double TOTAL_ROWS_COUNT = LEFT_ROWS_COUNT + RIGHT_ROWS_COUNT;
+    private static final double LEFT_JOIN_COLUMN_NULLS = 0.3;
+    private static final double LEFT_JOIN_COLUMN_NON_NULLS = 0.7;
+    private static final int LEFT_JOIN_COLUMN_NDV = 20;
+    private static final double RIGHT_JOIN_COLUMN_NULLS = 0.6;
+    private static final double RIGHT_JOIN_COLUMN_NON_NULLS = 0.4;
+    private static final int RIGHT_JOIN_COLUMN_NDV = 15;
+
+    private static final SymbolStatistics LEFT_OTHER_COLUMN_STATS =
+            symbolStatistics(LEFT_OTHER_COLUMN, 42, 42, 0.42, 42);
+    private static final SymbolStatistics RIGHT_OTHER_COLUMN_STATS =
+            symbolStatistics(RIGHT_OTHER_COLUMN, 24, 24, 0.24, 24);
+    private static final PlanNodeStatsEstimate LEFT_STATS = planNodeStats(LEFT_ROWS_COUNT,
+            symbolStatistics(LEFT_JOIN_COLUMN, 0.0, 10.0, LEFT_JOIN_COLUMN_NULLS, LEFT_JOIN_COLUMN_NDV),
+            LEFT_OTHER_COLUMN_STATS);
+    private static final PlanNodeStatsEstimate RIGHT_STATS = planNodeStats(RIGHT_ROWS_COUNT,
+            symbolStatistics(RIGHT_JOIN_COLUMN, 5.0, 10.0, RIGHT_JOIN_COLUMN_NULLS, RIGHT_JOIN_COLUMN_NDV),
+            RIGHT_OTHER_COLUMN_STATS);
+
+    private static final JoinStatsRule JOIN_STATS_RULE = new JoinStatsRule(new FilterStatsCalculator(createTestMetadataManager()));
+
+    private StatsCalculatorTester tester;
+
+    @BeforeMethod
+    public void setUp()
+            throws Exception
+    {
+        tester = new StatsCalculatorTester();
+    }
+
+    @Test
+    public void testStatsForInnerJoin()
+            throws Exception
+    {
+        double innerJoinRowCount = LEFT_ROWS_COUNT * RIGHT_ROWS_COUNT / LEFT_JOIN_COLUMN_NDV * LEFT_JOIN_COLUMN_NON_NULLS * RIGHT_JOIN_COLUMN_NON_NULLS;
```

Comment:
true, but the goal is to test that the implementation of the formula actually works and not the formula itself (we assume the formula makes sense). I know that those tests will be hard to maintain, but without some randomized data generator (according to our assumptions), I don't see how this can be improved.